File: src/tests/idl/resources/repo/cpp/counter.main.cpp
// apps
#include "counter.app.example.h"

#ifdef _WIN32
#include <windows.h>

void mysleep() { Sleep(3000); }
#else
#include <unistd.h>

void mysleep() { sleep(3); }
#endif

void dsn_app_registration_counter()
{
    // register all possible service apps
    dsn::service_app::register_factory<::dsn::example::counter_server_app>("server");
    dsn::service_app::register_factory<::dsn::example::counter_client_app>("client");
    dsn::service_app::register_factory<::dsn::example::counter_perf_test_client_app>("client.perf.counter");
}

#ifndef DSN_RUN_USE_SVCHOST

int main(int argc, char **argv)
{
    dsn_app_registration_counter();

    dsn_run_config(argv[1], false);
    mysleep();
    dsn_exit(0);
}

#else

#include <dsn/internal/module_int.cpp.h>

MODULE_INIT_BEGIN
dsn_app_registration_counter();
MODULE_INIT_END

#endif

File: src/tests/idl/resources/repo/csharp/counter.main.cs
using System;
using System.IO;
using System.Linq;
using System.Threading;
using System.Collections.Generic;

using dsn.dev.csharp;

namespace dsn.example 
{
    class Program
    {
        static void Main(string[] args)
        {
            counterHelper.InitCodes();
            testHelper.init();

            ServiceApp.RegisterApp<counterServerApp>("server");
            ServiceApp.RegisterApp<counterClientApp>("client");        

            string[] args2 = (new string[] { "counter" }).Union(args).ToArray();
                Native.dsn_run(args2.Length, args2, false);
           // Native.dsn_run_config("config.ini", false);
            int counter = 0;
            while (testHelper.finished_test() != 4)
            {
                Thread.Sleep(1000);
                if (++counter > 10)
                {
                    Console.WriteLine("it takes too long to complete...");
                    Native.dsn_exit(1);
                }
            }
            bool passed = true;
            Dictionary<string, bool> result = testHelper.result();
            foreach (KeyValuePair<string, bool> entry in result)
            {
                string name = entry.Key;
                bool ok = entry.Value;
                passed = passed && ok;
                Console.WriteLine(name + " " + (ok ? "passed" : "failed"));
            }
            int ret = passed ? 0 : 1;
            Native.dsn_exit(ret);
        }
    }
}
